### ISSUE TITLE
Add search functionality to rope

### ISSUE THREAD

**Issue by kevinjohnston** [kevinjohnston/doe-core#6]
**Goal**
Return a lazy sequence of all instances matching some regex in a given rope.

**Reason**
Using normal search functionality is impractical for large strings

**Thoughts:**
* Implement this after most rope optimizations. For now calling text and searching that is good enough
* Rope leaf nodes can searched in parallel
* If rope leafs contain an unrealized deferred, they should not be realized for this operation unless the lazy find result is evaluated. (to prevent memory issues with large strings becoming realized)

**See Also:**
